src/SNLS_device_forall.h

```diff
@@ -50,10 +56,99 @@ end,                                             \
 
 namespace snls {
 
-   typedef RAJA::View<bool, RAJA::Layout<1> > rview1b;
-   typedef RAJA::View<double, RAJA::Layout<1> > rview1d;
-   typedef RAJA::View<double, RAJA::Layout<2> > rview2d;
-   typedef RAJA::View<double, RAJA::Layout<3> > rview3d;
+   // Provide some simple shortcuts in-case people need something beyond the default
+   template<typename T>
+   using rview1 = RAJA::View<T, RAJA::Layout<1>>;
+   template<typename T>
+   using rview2 = RAJA::View<T, RAJA::Layout<2>>;
+   template<typename T>
+   using rview3 = RAJA::View<T, RAJA::Layout<3>>;
+
+   using rview1b = rview1<bool>;
+   using rview1d = rview1<double>;
+   using rview2d = rview2<double>;
+   using rview3d = rview3<double>;
+   using crview1d = rview1<const double>;
+   using crview2d = rview2<const double>;
+   using crview3d = rview3<const double>;
+
+   using rhost_res = RAJA::resources::Host;
+#if defined(__snls_gpu_active__)
+#if defined(RAJA_ENABLE_CUDA)
+         using rgpu_res = RAJA::resources::Cuda;
+#else //defined(RAJA_ENABLE_HIP)
+         using rgpu_res = RAJA::resources::Hip;
+#endif
+#endif
+
+   using rres = std::variant<
+      rhost_res
+#if defined(__snls_gpu_active__)
+      ,
+      rgpu_res
+#endif
+   >;
+
+   using rrese = RAJA::resources::Event;
+
+   // We really don't care what View class we're using as the sub-view just wraps it up
+   // and then allows us to take a slice/window of the original view
+   // Should probably work out some form of SFINAE to ensure T that we're templated on
+   // is an actual View that we can use.
+   template<class T>
+   class subview {
+   public:
+      // Delete the default constructor as that wouldn't be a valid object
+      __snls_hdev__
+      subview() = delete;
+      // where we don't want any offset within the subview itself
+      __snls_hdev__
+      subview(const int index, T& view) : m_view(view), m_index(index), m_offset(0) {};
```

Comment:
Should this be `const T& view` instead of `T& view` passed into the constructor? Same for the other constructor below.